3.13/redundancy_eliminator.md

```diff
@@ -0,0 +1,126 @@
+shall be an interpretation over Python symbolic information state,
+represented by `_Py_UOpsSymType *`.
+
+The abstract interepeter has the equivalent constructs of the
+CPython interpreter to represent state. It has symbolic
+local and symbolic frames, This allows it to emulate CPython
+runtime state accurately.
+
+Executing "steps" in the abstract interpreter is thus the same as
+doing the same in CPython -- by interpreting bytecode instructions.
+The abstract interpreter thus executes a bytecode, update its
+symbolic state, and repeats until it reaches a terminator (the end
+of the trace).
+
+### Lattice
+
+**Note: if you are uninterested in the theory behind data-flow analysis,\
+please skip this section.**
+
+The states of information in such an abstract interpretation can be
+represented by a [lattice](https://en.wikipedia.org/wiki/Lattice_(order))
+of values. The lattice itself is an ordered powerset
+of all variables and their associated information.
+
+For example, here is a trivial lattice for the statement 
+```python
+x = 1
+y = 2.0
+```
+:
+
+```mermaid
+flowchart TD
+    A("`Bottom, ie. x: *?*, y: *?*`") --> B("`x: *int*, y: *?*`")
+    A --> C("`x: *?*, y: *float*`")
+    B --> D("`Top, x: *int*, y: *float*`")
+    C --> D
+  
+```
+
+Real Python runtime information
+may contain more complex types, constants, and unions of types.
+
+Each `_Py_UOpsSymType` captures all the information associated
+with a variable at any point in the lattice.
+
+For more information regarding lattices in compiler theory,
+Page 27 onwards of this presentation provides a gentle introduction
+https://ilyasergey.net/CS4212/_static/lectures/PLDI-Week-12-dataflow.pdf
+Credits to the courses UPenn Compilers (CIS 341) and NUS Compiler
+Design (CS4212).
+
+### Abstract DSL
+
+A extra specification that overrides the original cases in
+`Python/bytecodes.c` is used.
+This generates an abstract interpreter that operates on these symbolic values.
+Where there is no overridden case, the generator falls back to a generic
+version where all output stack values are unknown symbolic values.
+
+```
+// Python/bytecodes.c
+op(OP, (arg1 -- out)) {
+    eggs();
+}
+```
+```
+// Python/tier2_redundancy_eliminator_bytecodes.c
+// Nothing to see here :).
+```
+```
+// Python/abstract_interp_cases.c.h
+case OP: {
+    _Py_UOpsSymType *out;
+    out = sym_init_unknown(ctx);
+    if (out == NULL) goto error;
+    stack_pointer[0] = out;
+    stack_pointer += 1;
+    break;
+}
+```
+
+Where this is an overridden case, all information from the overridder
+takes precedence. For example:
+
+```
+// Python/bytecodes.c
+op(OP, (arg1 -- out)) {
+    eggs();
+}
+```
+```
+// Python/tier2_redundancy_eliminator_bytecodes.c
+op(OP, (arg1 -- out)) {
+    spam();
+}
+```
+```
+// Python/abstract_interp_cases.c.h
+case OP: {
+    _Py_UOpsSymType *arg1;
+    _Py_UOpsSymType *out;
+    arg1 = stack_pointer[-1];
+    spam();
```

Comment:
```suggestion
    out = spam(arg1);
```
(Note that the override *must* set `out`, otherwise we'd put an uninitialized pointer on top of the stack.)